File: google-scholar-scraper/api/python/request.py
"""Google Scholar via the Scrapeless Scraper API (scraper.google.scholar).

This actor is synchronous but FLAKY: it intermittently returns
{"code": 20500, "message": "scraping failed"}. Retry until the body has
"scholar_result".

Run:
    export SCRAPELESS_API_KEY=your_api_token_here
    python request.py
"""
import os
import json
import time
import requests

ENDPOINT = "https://api.scrapeless.com/api/v1/scraper/request"


def scrape_scholar(query: str, hl: str = "en", max_attempts: int = 6) -> dict:
    headers = {
        "Content-Type": "application/json",
        "x-api-token": os.environ["SCRAPELESS_API_KEY"],
    }
    payload = {"actor": "scraper.google.scholar", "input": {"q": query, "hl": hl}}
    for attempt in range(1, max_attempts + 1):
        resp = requests.post(ENDPOINT, headers=headers, json=payload, timeout=120)
        resp.raise_for_status()
        data = resp.json()
        # Success looks like {"metadata": ..., "scholar_result": ...}.
        if "scholar_result" in data:
            return data
        # Otherwise it's a transient {"code": 20500, ...}; retry.
        print(f"attempt {attempt}/{max_attempts} failed ({data}); retrying…")
        time.sleep(3)
    raise RuntimeError(f"scraper.google.scholar still failing after {max_attempts} attempts")


if __name__ == "__main__":
    data = scrape_scholar("transformer neural network")
    # The parsed academic results live under data["scholar_result"].
    print(json.dumps(data.get("scholar_result", data), indent=2, ensure_ascii=False))

File: google-scholar-scraper/api/nodejs/request.mjs
// Google Scholar via the Scrapeless Scraper API (scraper.google.scholar).
// This actor is synchronous but FLAKY: it intermittently returns
// {"code":20500,"message":"scraping failed"}. Retry until the body has scholar_result.
//   export SCRAPELESS_API_KEY=your_api_token_here
//   node request.mjs
const ENDPOINT = "https://api.scrapeless.com/api/v1/scraper/request";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

export async function scrapeScholar(query, { hl = "en", maxAttempts = 6 } = {}) {
  const headers = {
    "Content-Type": "application/json",
    "x-api-token": process.env.SCRAPELESS_API_KEY,
  };
  const body = JSON.stringify({
    actor: "scraper.google.scholar",
    input: { q: query, hl },
  });
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    const resp = await fetch(ENDPOINT, { method: "POST", headers, body });
    if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${await resp.text()}`);
    const data = await resp.json();
    // Success looks like { metadata, scholar_result }.
    if (data.scholar_result) return data;
    // Otherwise it's a transient { code: 20500, ... }; retry.
    console.error(`attempt ${attempt}/${maxAttempts} failed (${JSON.stringify(data)}); retrying…`);
    await sleep(3000);
  }
  throw new Error(`scraper.google.scholar still failing after ${maxAttempts} attempts`);
}

const data = await scrapeScholar("transformer neural network");
// The parsed academic results live under data.scholar_result.
console.log(JSON.stringify(data.scholar_result ?? data, null, 2));
